### DISCUSSION TITLE
Proposal: Vocabularies are Always Mandatory

### DISCUSSION THREAD

**Discussion by gregsdennis** [json-schema-org/community#342] [Specification]
Previously, on JSON Schema Discussions:

- Dropping support for unknown keywords was [proposed](https://github.com/orgs/json-schema-org/discussions/241) and [mostly accepted](https://json-schema.org/blog/posts/the-last-breaking-change)
- Unknown keywords are still going to be [kinda supported](https://github.com/orgs/json-schema-org/discussions/329)
- Other stuff that's irrelevant to this discussion

Currently, a vocabulary is listed in a meta-schema by URI ID along with a boolean value that indicates the required behavior for an implementation that **_does not_** understand the vocabulary:

- If the value for the vocabulary is `true`, then the implementation MUST refuse to process the schema.
- If the value for the vocabulary is `false`, then the implementation MAY continue to process the schema.

The second case, `false`, means the vocabulary is considered "optional" in that it's not required that the implementation understand the vocabulary.

However, now that we've said unknown keywords are disallowed, we have a conundrum:

The keywords in an unknown vocabulary are themselves unknown, and implementations MUST refuse to process them.  But this means that implementations no longer have the option to continue processing.  Therefore the `false` case now indirectly requires the same behavior as the `true` case.

Thus, **if implementations must refuse to process schemas with unknown keywords, a vocabulary can't really be optional,** and the true/false values are meaningless.

---

The fallout from this is that the syntax of the `$vocabulary` keyword

```
{
  "$vocabulary": {
    "https://my.schema.com/vocabs/myvocab": true, 
    ...
  }
}
```

doesn't really make sense anymore.

Should we change the keyword to just an array of URI IDs?  Do something else?

The format of this keyword was never really solidified as vocabularies as a feature are still being developed, so I don't mind keeping its definition a bit fluid for now.

**Comment by jdesrosiers** [comment]
Whether optional vocabularies make sense or not depends on what "optional" means. I think we can define it a way where it makes sense, even if it's not quite as useful as it used to be. I would define an "optional" vocabulary to mean that implementations for the keywords in that library are optional. An implementation would need to "know" all vocabularies in `$vocabularies`, but doesn't need to "implement" keywords defined in optional vocabularies.

For example, my implementation doesn't support the "format-assertion" version of `format`. It knows that the "format-assertion" vocabulary defines the `format` keyword, but doesn't implement that keyword. If the dialect has `".../vocab/format-assertion": false` and the schema has the `format` keyword, I can still evaluate that schema. However, if the dialect was `".../vocab/format-assertion": true`, I wouldn't be able to evaluate the schema. In the future, if we add a vocabulary description format that includes what keywords the vocabulary defines, implementations could again support optional vocabularies they have no code for by looking up the vocabulary description.

This discussion is about reconciling optional vocabularies with forbidding unknown keywords. I think this approach does that in the least intrusive way. However, I think there's a less urgent discussion we should probably have about whether we think it even makes sense for the vocabulary system to have a concept of optional vocabularies.

**Reply by gregsdennis** [reply]
>  If the dialect has ".../vocab/format-assertion": false and the schema has the format keyword, I can still evaluate that schema.

True, but you'd likely get different validation results from an implementation that _does_ process that keyword with no indication as to why.  Is there value in that?

[This text](https://json-schema.org/draft/2020-12/json-schema-core.html#section-8.1.2-4) would need to change:

> The values of the object properties MUST be booleans. If the value is true, then implementations that do not recognize the vocabulary MUST refuse to process any schemas that declare this meta-schema with "$schema". If the value is false, implementations that do not recognize the vocabulary SHOULD proceed with processing such schemas. The value has no impact if the implementation understands the vocabulary.

Implementations would no longer be able to process schemas that use vocabs they don't "know."  They'd have to know all vocabs to proceed, but they aren't required to be able to process the keywords.

Yeah, that just doesn't sound like a good idea to me.  It's encouraging the hack of "Well, the lib doesn't support this vocab, so we'll fudge it by fake-implementing it.  The implementation will think it knows the vocab, but won't process anything."  People are probably going to do that anyway, even with _required_ vocabs, but why give them a way to do it?  Why make it okay?

**Reply by Relequestual** [reply]
I mostly agree with @jdesrosiers's interpritation/suggestion of the desired way to understand "known" and "unknown".
I agree with @gregsdennis's suggested required text change to enable that.

Initially, I was thinking we should continue to allow non-mandatory vocabularies, then reading this, I changed my mind... and back again.

> ...True, but you'd likely get different validation results from an implementation that does process that keyword with no indication as to why. Is there value in that?

I agree with this pretty strongly. One mantra I alwasy come back to with JSON Schema is "no surprises".

In terms of vocabularies which don't impact validation, I think it would be enough for the vocabulary meta-schema to be provided, and "known" or "unknown" keywords be determined based on that (Although, I recognise that we have continually stated that the meta-schemas are informative only).

Having an optional vocabulary which could impact validaiton results, breaks the "no surprises" rule I hold.

However there's at least one use case that comes to mind immediatly.

Say there's a vocabulary which checks that IDs are valid and found in the database.
You don't want your API gateway to be checking those, but you may want your application layer to check them and handle the erorr accordingly.

How could we avoid surprises?

Rather than just "true/false", we could allow for additional meta-data.
For example, if the value was an object rather than false, it could define the cases in which the vocabulary is required / not required. The cases could be user defined.
Implementations would need to support being provided with a "case" in which they are being used, and throw an error if they encounter a vocabulary which is required to be understood and processed for their provided case.
(You could have pre-processing tools which modify the meta-schema based on a given case, but I'm not totally sure how that would work, nor if it's actually a good or valid idea).

The truth is, most people won't be needing this. But for those that do, such a solution feels like a small detail that would be easy to understand, and result in no surprises.

**Reply by gregsdennis** [reply]
> In terms of vocabularies which don't impact validation

This is the key.  A vocabulary could only be optional if it doesn't impact your use case.  For example, a codegen vocab could be optional in a validation context, but when you get to the codegen context, it should be required.  What you really need is two meta-schemas: one for validation that doesn't have the codegen vocab, and another for codegen that does have the vocab (and it's required).

Then, on top of that, you have to have two implementations: one for each context.  If you're going to have a single implementation, then it's going to need to understand the codegen vocab anyway, which means that you might as well require the vocab since it's required in one of your contexts, meaning there's no need for the vocab to be optional.

> Rather than just "true/false", we could allow for additional meta-data.

I don't know how we'd specify this to be interoperable.  That seems like a lot of work for getting this next spec out.  Maybe for the future.

**Reply by jdesrosiers** [reply]
I think that what we do with optional vocabularies (and the vocabulary system in general) is a bigger conversation and a low priority at this point. All of the arguments for why not to support optional vocabularies apply equally with or without allowing unknown keywords, so I don't think this is something that is blocking making that change. The change I suggested unblocks us without requiring a bigger discussion about an overhaul of the vocabulary system. It's definitely a discussion that needs to happen, but I think it's something we can tackle after the next release.

**Reply by gregsdennis** [reply]
I thought so, too, until I started making the changes.  The problem is that an unknown optional vocabulary produces unknown keywords, which are disallowed.

The only case where it doesn't is if someone explicitly writes code to "know" but not support a vocabulary, which

- produces inconsistent results (in general, "surprises")
- can still be done for required vocabs

I think this needs to be resolved before we can move on.

**Reply by jdesrosiers** [reply]
I think there's a miscommunication here. I'm not saying that there's not a problem here that needs to be addressed. I'm saying that we can address that problem without having to decide whether we should support optional vocabularies. With just a minor adjustment to the definition of what it means for a vocabulary to be "optional", we can avoid opening a huge can of worms.

Just so we're on the same page, setting aside the question of whether or not the concept of optional vocabularies is a good idea, and setting aside whether or not you like the solution I presented, do you agree that the solution I presented is a potential solution to the problem, or are you trying to tell me that my solution _can't_ work?

**Reply by gregsdennis** [reply]
> I think there's a miscommunication here. I'm not saying that there's not a problem here that needs to be addressed. I'm saying that we can address that problem without having to decide whether we should support optional vocabularies.

What's the problem you see that needs addressing?

For me, the problem (and the reason for opening the discussion) _is_ that optional vocabularies have no utility and therefore should not be supported.

**Reply by SorinGFS** [reply]
@gregsdennis 
As you [said here](https://github.com/orgs/json-schema-org/discussions/370#discussioncomment-5525768) `$vocabulary` is still a work in progress having multiple uncertainties. I think you should postpone the next release until all these uncertainties are resolved, or exclude the `$vocabulary` keyword from the next release.

**Reply by gregsdennis** [reply]
@SorinGFS thank you for your opinion, but that is part of a broader conversation on whether to include experimental features.  It is off-topic for this discussion.

Please see Jason's [SDLC proposal](https://github.com/orgs/json-schema-org/discussions/234) for a proposal on how experimental features could be included in the future.  That conversation has been put on hold while we nail down some more immediate concerns.

**Comment by gregsdennis** [comment]
It's worth pointing out that the [current document](https://github.com/json-schema-org/json-schema-spec/blob/b11e1dc9831711ac2e4af6b70d6861c94aa9de6f/jsonschema-core.md?plain=1#L958-L959) (in GitHub) also recognizes this relationship:

> Implementations that do not support a vocabulary that is optionally used by a
> schema SHOULD proceed with processing the schema. **The keywords will be
> considered to be unrecognized keywords as addressed by {{unrecognized}}.**

The direct consequence of this is that if unsupported optional vocabularies create unrecognized keywords, and unrecognized keywords are forbidden, then optional vocabularies are actually required.

**Comment by karenetheridge** [comment]
I think the only way out here (to continue allowing a vocabulary to be specified as `false`) would be to explicitly list the keywords that such a vocabulary defines, so they can be treated as "ignored" (or annotation-only) rather than "error on first sight".

The other useful usecase for true/false vocabularies is where two vocabularies define the same set of keywords, and the schema has the option of deciding which to support - e.g. the two format vocabularies which both define `format` but in different ways. Because the format-annotation vocabulary is mandatory, we know that implementations will not treat the `format` keyword as unknown.

**Comment by gregsdennis** [comment]
I would like to point out that @Relequestual also (and independently) [noticed this conflict](https://github.com/json-schema-org/json-schema-spec/pull/1512#discussion_r1644182009) when attempting to remove support for unknown keywords.

**Reply by jdesrosiers** [reply]
He's also expressed discomfort with dropping the concept of optional vocabularies. There's no question that there's a conflict that needs to be addressed. I still think there's a path forward that makes sense and allows for the concept of optional vocabularies to still exist.
